docs/cardano-monads.rst

```diff
@@ -0,0 +1,467 @@
+  stacks. (As discovered later, this can be addressed with closed effect sums,
+  but then we lose extensibility, so it's still an issue).
+
+* Cost of introducing new effects. For ``N`` transformers and ``M`` classes,
+  there must be ``N * M - k`` instances (where ``k`` is the amount of impossible
+  effect combinations: one does not simply use ``WriterT`` with ``ContT``). The
+  more effects we have in the application, the more costly it is to add one more
+  (at 10 effects we already hit 100 instances, the cost is insurmountable). Not
+  only this is boilerplate-heavy, it's also antimodular, as every effect needs
+  to know about each other.
+
+Conclusion:
+
+* In retrospect, we could probably solve the ``N*M`` issue by careful use of
+  ``{-# OVERLAPPABLE #-}`` instances (although ``Mockable``-related type
+  families would still lead to a fairly large amount of boilerplate).
+  Nevertheless, the run-time performance of highly nested transformer stacks
+  might be unsatisfactory. (Slowdown is linear in the amount of layers).
+
+
+The Bronze Age
+~~~~~~~~~~~~~~~
+
+The core observation is that many domain-specific effects are isomorphic to
+general purpose ones. The majority of monad transformers in Cardano SL were
+basically clones of ``ReaderT`` or ``StateT``. This is why we decided to use
+``ether``: unlike ``mtl``, it allows using multiple transformers of the same
+sort (i.e. multiple ``ReaderT``) in the same transformer stack, so we could
+simply replace our hand-crafted transformers with those from ``ether`` and put
+an end to the ``N*M`` problem. Furthermore, ``ether`` offers *flattening*,
+solving the run-time performance issue as well.
+
+Typically, the code would look like this::
+
+    type MonadFoo = Ether.MonadReader Foo
+    type FooT = Ether.ReaderT Foo
+
+    type MonadBar = Ether.MonadReader Bar
+    type BarT = Ether.ReaderT Bar
+
+Even when we needed a custom class with methods (rather than a synonym for
+``Ether.MonadReader``), we could define a single instance for ``IdentityT``, no
+boilerplate required.
+
+We used flattening to achieve good run-time performance, but kept many custom
+classes with effect operations (to allow for varying implementations). This led
+to a peculiar situation where the transformer stack consisted mostly of
+``IdentityT``::
+
+    type M =
+        TaggedTrans FooEff IdentityT $
+        TaggedTrans BarEff IdentityT $
+        TaggedTrans BazEff IdentityT $
+        ReaderT (FooEnv, BarEnv, BazEnv) IO
+
+Pros:
+
+* Extensibility. Introducing a new effect is really cheap. The code is modular
+  and effects don't need to know about each other.
+
+* Flexibility. (Same as above, local use of effects).
+
+* Good run-time performance. Since in the end the entire monad transformer stack
+  was just ``ReaderT`` with a bunch of ``IdentityT`` on top (and occasional
+  local ``StateT``), we enjoyed good run-time performance.
+
+* Conciseness. No boilerplate.
+
+Cons:
```

Comment:
You didn't describe how redirects had worked btw (which had been kind of messy).

Reply:
I don't think it was messy.

Reply:
`k + 2` transformer layers instead of `ReaderT IO`? It is for sure, at least some people from the team will agree with me, I believe.

Reply:
I don't understand your point. Okay, you can say that `k + 2` transformer layers instead of `ReaderT IO` is more complex than just `ReaderT IO`. But these transformer layers were specified in a single place (where the application was run), and they were specifying the implementation used for effects. And with `ReaderT IO` you have to specify these implementations somewhere as well, just not using transformers, so the complexity is still there, just in a different form.

Reply:
>

Reply:
With `ReaderT` implementation is explicitly defined through the set of instance declarations that we put somewhere nearby. With redirects, it's the same (we define the stack, we write instances), but now stack has `k` extra layers. 

There's no point, it's just a smol talk. :)